snapshots/windows/windows.go

```diff
@@ -312,12 +314,14 @@ func (s *snapshotter) mounts(sn storage.Snapshot) []mount.Mount {
 		roFlag = "rw"
 	}
 
-	if len(sn.ParentIDs) == 0 || sn.Kind == snapshots.KindActive {
-		source = s.getSnapshotDir(sn.ID)
-		parentLayerPaths = s.parentIDsToParentPaths(sn.ParentIDs)
-	} else {
-		source = s.getSnapshotDir(sn.ParentIDs[0])
-		parentLayerPaths = s.parentIDsToParentPaths(sn.ParentIDs[1:])
+	source := s.getSnapshotDir(sn.ID)
+	parentLayerPaths := s.parentIDsToParentPaths(sn.ParentIDs)
+
+	mountType := "windows-layer"
+
+	if len(sn.ParentIDs) == 0 {
+		// A mount of a parentless snapshot is a bind-mount.
+		mountType = "bind"
```

Comment:
Because the mount code does the `Files/` sub-directory thing, is there any chance that calling this a `bind` mount (which seemed cute when I first faked it with symlinks) is going to confuse anything/anyone? Something like `windows-base-layer` would be nice and distinct.

I wonder if we could just use `windows-layer` for this too and count the number of parent layers instead? The only time we distinguish the two mount-types, we also have a call to m.GetParentPaths or access to its count, and one of those times, we're erroring for the "bind with parents" and "non-bind without parents" cases anyway. So it _seems_ like it would work, on inspection.

Reply:
I am inclined to leave this as a ```bind``` mount. In effect, it walks and it talks like a bind mount, and it is far more intuitive to understand to anyone coming from the *nix world than a special case of ```windows-layer```. Coincidentally, the filter we use to make this happen is also called ```BindFilter``` :smile:.

I don't feel particularly strongly about this, so if you feel like this is something that needs to be changed, I will make the changes.

Reply:
That was actually why I was wondering if it should be changed, because it's _close_ to a Linux bind mount, but by handling `Files/` the way we do now, I was wondering if that was a false equivalence, since we currently assume/expect that the mount source being passed to it is a WCOW base layer, not an arbitrary directory.

That would allow us to reserve `bind` for an _actual_ arbitrary directory bind-mount that a caller might request with a hand-constructed `Mount` object, which _is_ what we had implemented before ebf9906ed6336c95ba9cc6be74d83ce858e30958, on reflection.

So maybe ebf9906ed6336c95ba9cc6be74d83ce858e30958 was wrong-headed, and a base-layer-only Active mount _should_ be a trivial bind-mount to `Files/`, with a special case for unpacking which returns a mount which will do the wrong thing if mounted? The special case for unpacking in `createSnapshot` already "returns a mount which will do the wrong thing if mounted" for non-base layers, so this isn't as bad as my phrasing makes it sound; it'd be consistent, at least. I had a quick poke around with GitHub, and the only places in containerd that use the unpacking key are local operations that immediately `archive.Apply` and `snapshot.Commit`, so the only place this has risk is users directly calling the snapshotter API.

Anyway, you're closer to working with this code than I am now, so I'm fine with whatever you feel minimises surprise/risk etc. My feeling is that the current state is caught between two worlds, and we should either move the append of `Files/` back into the snapshotter (restoring "Bind mount does what it says on the box") or we push through with "`windows-filter` is the only mount-type we know, and it knows how to handle both base and non-base layers".

I also feel that if we're going to go with "bind", such mounts should not have `parentLayersOption` included in the `Mount` object, since by snapshotter definition that's empty for a bind mount, and if it's not empty, we'd reject it at mount-time anyway. (The presence of `parentLayersOption` in bind mounts is what opened this question up for me in the first place.)

Reply:
I'll have to think about this for a bit.

The ability to "bind" mount is already (sort of) used in ```buildkit``` here: https://github.com/moby/buildkit/blob/master/snapshot/localmounter_windows.go#L30-L52

Some clients already expect to be able to "bind" mount and have that work in a certain way. If we decide to do away with "bind" mounts now, until we can decide on a more accurate implementation, the ```buildkit``` code can be adapted to Windows, but would introduce additional code paths to account for the differences.

The way Windows layers work make it a bit tricky. From a client's perspective, the ```Files``` folder is the actual "```rootfs```". If I understand things correctly, even for a ```Hyper-V``` isolated container (Xenon), the same rule applies. The difference is, we also have the ```UtilityVM``` with it's own files needed to boot up a VM, in which the actual container is started, and for which the ```Files``` folder becomes the ```rootfs``` of the container. Sort of like Kata containers if we think about it, but in this case, the plumbing that Kata adds is included in the actual layer on Windows.

So the code as it is now, enables clients to mutate or view the ```rootfs``` of a container, completely ignoring any support files the layer contains. The same thing happens on *nix systems, as those don't really have all the supporting files for ```UtilityVMs```, overlay hives or the needed BCD hives to actually "boot" a ```UtilityVM``` or a container.

As far as the clients are concerned, the API is similar to what they get on Linux. If I understand your concerns correctly, there remains the issue where we want to mount a layer for the purpose of unpacking an actual image into it. I will have to follow that code on Monday and see how that happens. I haven't looked at that much.

As for clients, I am trying to think of a reason a client might want to bind mount the actual layer, as opposed to the "```rootfs```" of the containers that resides in that layer. Mounting a ```windows-layer``` on a host will also expose the ```Files``` folder. So it doesn't seem like clients should be concerned with the internal plumbing of how a Windows container layer is handled. Further more, I would be inclined to prevent direct access to the base layer for clients, as there is no real reason to mess with internals as a client and that might actually break things in unexpected ways, if the "support" files are tampered with.

For now, I will revert this PR pre https://github.com/containerd/containerd/commit/ebf9906ed6336c95ba9cc6be74d83ce858e30958. If you're willing, we can have a broader discussion on slack, at your convenience, sometime next week. But only if you can spare ~1 hour or so. No worries if not.

Reply:
> I also feel that if we're going to go with "bind", such mounts should not have `parentLayersOption` included in the `Mount` object, since by snapshotter definition that's empty for a bind mount, and if it's not empty, we'd reject it at mount-time anyway.

I've added: https://github.com/containerd/containerd/pull/8043/commits/ca5605b4a208d9d5c809813adbec6b2227ab6c23

Reply:
I haven't gone through the rest (on my way to bed) but...

> The ability to "bind" mount is already (sort of) used in `buildkit` here: https://github.com/moby/buildkit/blob/master/snapshot/localmounter_windows.go#L30-L52

I'll note that I'm partly responsible for that code, and it contains at least one complete misunderstanding by me of how WCOW layer mounts work, right at the start of this adventure. (But not the bit you're talking about here, that's common with the Linux localmounter)

It does emphasise my point somewhat that if we say "bind" as the type, then `Mount.source` _better_ be what will appear at the target location, since BuildKit does this "Bypass rw bind mounts" trick. What we have now will break that, and if we revert to the state we were in before (which on skimming, I see you did), then we'll be in a better place.

I _think_ that once we have landed this (whether we use "bind" and BuildKit's bypass works, or "windows-filter" and BuildKit trustingly calls `mount.All` to create a sparkling symlink), the BuildKit localmounter code for non-Windows should also be usable on Windows allowing us to unify that split. It's possible `MNT_DETACH` will bite us here though, I don't recall if `mount.Unmount` ignores options on Windows or enforces 0.

----

Okay, I did read the rest of the comment, and generally agree with you. I will note that for non-base layers (`windows-layer`) mounts, the `Mount.Source` also has all the metadata etc. visible. So if we call the base layers `windows-layer`, then `Mount.Source` exposing the metadata is consistent. I agree that if we call it a `bind` mount, we should not expose the metadata, just point at the contained files, like you'd see if you actually mounted it.

The other user of the Mount structure outside of containerd is the runtime shims; hcsshim of course needs to see the metadata, to launch a container, but we'll never give it a base-layer mount as it cannot execute that (it needs a sandbox on top, AFAIK, although Host Process Containers work under very different limitations).

The most-conservative route would be to only return `windows-layer` mounts (i.e. the state of the world before this PR), and hence even though a base-layer mount would happen to be a trivial bind-mount, inspecting the Mount object wouldn't tell you that. This would let us reduce the "unpack" behaviour differences, and prevents the BuildKit "rw-bind-mount" bypass.

However, if you're confident that for a base-layer snapshot, we can produce a Mount object that looks just like a bind-mount on Linux, then that's not an issue. That's the state I think we were in before https://github.com/containerd/containerd/pull/8043#discussion_r1153930966, and at this point, I don't have any cases where it'll go wronger than the existing code that skips sandbox.vhdx creation, and your testing hadn't shown up any issues with BuildKit here.

Reply:
Apologies for the wall of text :smile:

I've been thinking about the issue of mounts and how we could clarify what we want to achieve here. A bit of context first, so we don't get lost in details. You are aware of most of this context, as you've basically written this :smile:, but for the sake of keeping the conversation on track, I think it's worth adding here.

## Arbitrary base layers

This feature came (mostly) out of the need for buildkit to be able to [create mutable refs when it syncs "local" sources](https://github.com/moby/buildkit/blob/master/source/local/local.go#L127), but also to enable the ability of generating layers with arbitrary files that could be exported and imported again. These layers are not necessarily usable as layers from which we can "boot" a container.

This work has a hcsshim component that creates a number of files needed for a layer to be considered valid. The layout of the layer requires that there is a ```Files``` folder, which becomes the ```rootfs``` of the container, if this layer were to be "bootable". Currently, ```hcsshim``` and ```HCS``` in general doesn't have any other kind of layer it supports, but ```WCIFS```, so we create a valid ```WCIFS``` layer to "trick" it.

## The containerd snapshotter

The code proposed here, leverages that new functionality in ```hcsshim``` and enables the creation of new, parent-less base layers. The code currently assigns a new mount type called ```bind``` for these parent-less layers, and mutates the source to the ```Files``` folder.

The client can then choose to mount this new ```mount.Mount{}``` at the desired location using ```m.Mount(target string)```.

## Mounting a ```mount.Mount{}```

### Windows layers

On Windows, before this PR, the only mount type supported was ```windows-layer```. Before a windows layer can be mounted on the host, we need to "activate" it, which creates a new ```\\?\Volume{}``` via ````WCIFS```. At least for writable layers. This new volume is a "gateway" to the ```Files``` folder. We can mount it on a host and mutate it if we need to.

For parent-less base layers, ```ActivateLayer()``` and ```PrepareLayer()``` don't really create a new WCIFS volume. Calling ```GetLayerMountPath()``` on an activated parent-less base layer, will return the actual snapshot location.

 ### The issue of "bind" mounts

Bind mounts, as they are introduced here, are a special case, and in the case of Windows, may be useful, but only in the context of a "local" mounter. The shim cannot handle "bind" mounts. So we have a choice to either "support" this in containerd when mounting locally, or just do away with it. We could in theory simply [leave the if statement](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR52-R57), and it should be fine, but that is about it in terms of "bind" support.

## Suggested course of action

Given the above, here is what I suggest:

  * The snapshotter should not concern itself with any kind of layers besides ```windows-layer```.
    * After creating a new parent-less layer, the [mounts()](https://github.com/containerd/containerd/pull/8043/files#diff-b6ae335f862c747466f1bc00e89eb1ef7cd9589bcc89ec57953c9c20125b3752R306) function should **not** concern itself with the mount type. It will remain a ```windows-layer```, but we will however check for the existence of parent layers and only add ```parentLayersOption``` if there are any parent layers.
    * We will no longer append ```Files``` to source, but simply return it as is.
  * In [mount.mount()](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR102), right before ```ApplyFileBinding()```, we check if this layer has any parents. In the case of a parent-less layer, [the volume](https://github.com/containerd/containerd/pull/8043/files#diff-ede639814a8abed289b5808723cc84b4aeee754dfc97fbaa435fb0dc4e910b7dR97) will not be a WCIFS ```\\?\Volume{}```, but the actual path on disk to the snapshot location. We cannot mount this, as a client like buildkit will potentially remove everything before writing something to the target. This, for example, actually happens due to how ```diffcopy``` works in fsutil. Source and target are synced something like how ```rsync -a --delete``` works.
    * In case of a WCIFS volume, we simply mount it, as it already is a gateway inside ```Files```
    * In the case of a parent-less base layer, we check for the existence of ```Files``` and mount that inside the target.

The local mounter in buildkit will be amended. Right now it mounts layers in-place, something that is not necessary. We also allow ```Mount()``` to do the right thing, instead of returning the source if it's not ```ro```.

What do you think?

Reply:
I've pushed: https://github.com/containerd/containerd/pull/8043/commits/d373ebc4de9a514af7d42c2edc980ec1d0868598

Reply:
Marking this as resolved. Let me know if you have any thoughts on the above or if you can think of anything I've missed.